Add auto-configuration for OpenTelemetry SdkMeterProvider

ThomasVitale opened the pull request:
- Introduced `metrics` sub-package in the `spring-boot-opentelemetry` module.
- Provided auto-configuration for `SdkMeterProvider` with customisation options via configuration properties and customizer bean.
- Defined missing OpenTelemetry `Clock` bean used to initialise observability signals providers.
- Added automated tests for the newly added auto-configuration.

Fixes gh-36546

ThomasVitale commented on module/spring-boot-opentelemetry/src/main/java/org/springframework/boot/opentelemetry/autoconfigure/metrics/OpenTelemetryMetricsAutoConfiguration.java:
```diff
@@ -0,0 +1,88 @@
+			ObjectProvider<SdkMeterProviderBuilderCustomizer> customizers) {
+		SdkMeterProviderBuilder builder = SdkMeterProvider.builder().setClock(clock).setResource(resource);
+		if (properties.getExemplars().isEnabled()) {
+			SdkMeterProviderUtil.setExemplarFilter(builder, exemplarFilter);
```
Exemplars are disabled by default because even though the Spec/API is GA, the method in the Java SDK to configure them is not exposed publicly, but through this util. Is that acceptable for Spring Boot? 

The API (mapped closely by the configuration properties) won't change as it's stable. So, any change to the SDK internals will not require any code/configuration change from Spring Boot users (a.k.a. stable feature). See: https://opentelemetry.io/docs/specs/otel/metrics/sdk/#exemplar (stable API). More context: https://github.com/open-telemetry/opentelemetry-java/issues/7503.

My recommendation would be to keep this functionality, but disabled by default. Considering that the OTLP Meter Registry doesn't support exemplars yet, this would be the only way in Spring Boot to support exemplars in OpenTelemetry.

mhalbritter commented:
Thanks for the PR @ThomasVitale . Before we merge this, we need to decide on https://github.com/spring-projects/spring-boot/issues/36546#issuecomment-3143885963, because this PR only provides some OTel beans but no way of exporting the metrics to a backend.

ThomasVitale commented on module/spring-boot-opentelemetry/src/test/java/org/springframework/boot/opentelemetry/autoconfigure/metrics/OpenTelemetryMetricsPropertiesTests.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Thomas Vitale
+ */
+class OpenTelemetryMetricsPropertiesTests {
```
I'll remove them

ThomasVitale commented on module/spring-boot-opentelemetry/src/main/java/org/springframework/boot/opentelemetry/autoconfigure/metrics/OpenTelemetryMetricsProperties.java:
```diff
@@ -0,0 +1,108 @@
+	/**
+	 * Maximum number of distinct points per metric.
+	 */
+	private Integer cardinalityLimit = 2000;
```
Done

ThomasVitale commented on module/spring-boot-opentelemetry/src/main/java/org/springframework/boot/opentelemetry/autoconfigure/metrics/OpenTelemetryMetricsAutoConfiguration.java:
```diff
@@ -0,0 +1,88 @@
+
+	@Bean
+	@ConditionalOnMissingBean
+	@ConditionalOnBean(OpenTelemetry.class)
```
Removed

ThomasVitale commented:
@mhalbritter thanks so much for the review. I have addressed your comments and pushed a second commit (after the final review, I can squash them to provide a single commit).

mhalbritter commented:
As explained [in this comment](https://github.com/spring-projects/spring-boot/issues/36546#issuecomment-3265547438), we won't add any auto-configuration for OpenTelemetry metrics. Thanks for your effort, anyway!
